### PR TITLE
s2255: getUser and getAugmented user function with multiple ldap entries returned.

### PR DESCRIPTION
**Pull request by hjeeves**

### CODE REVIEW

File: cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java [hunk 1]
```diff
@@ -525,61 +610,82 @@ private User getUser(final Principal userID, final String usersDN)
...
-            user.personalDetails.city = searchResult.getAttributeValue(LDAP_CITY);
-            user.personalDetails.country = searchResult.getAttributeValue(LDAP_COUNTRY);
-            user.personalDetails.email = searchResult.getAttributeValue(LDAP_EMAIL);
-            user.personalDetails.institute = searchResult.getAttributeValue(LDAP_INSTITUTE);
+            LdapDAO.checkLdapResult(e.getResultCode());
         }
 
-        logger.debug("getUser: found " + userID.getName() + " in " + usersDN);
-        return user;
+        logger.debug("getUserList returning " + userList.size() + " entries for " + userID.toString());
```
Comment (andamian) [inline cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java]: Is this the only place where multiple entries are logged? If so, shouldn't it be a higher level of debugging?

File: cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java [hunk 2]
```diff
@@ -351,6 +353,14 @@ private String getEmailAddress(final User user)
     private void checkUsers(final Principal userID, final String email, final String usersDN)
         throws TransientException, UserAlreadyExistsException
     {
+        // Internal system property set during LdapUserDAOTest.testAddMultipleUser
+        // so that this check can be overridden
+        if (System.getProperty(SUPPRESS_CHECKUSER_KEY) != null && System.getProperty(SUPPRESS_CHECKUSER_KEY).equals("true"))
+        {
```
Comment (andamian) [inline cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java]: Not a very big fan of this approach. Subclassing LdapUserDAO in the test code and overriding this method would have achieved the same effect without creating this backdoor.

File: cadc-access-control-server/src/test/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAOTest.java [hunk 3]
```diff
@@ -849,7 +852,155 @@ public Object run() throws Exception
...
+                    // Check to see if there are multiple users already existing.
+                    List<User> userList = userDAO.getAllUsers(userID, config.getUsersDN());
+
+                    if (userList.size() < 6) {
+                        log.debug("Creating multiple users (5 more)");
+
+                        // Run this 5x to create multiple users
+                        for (int i = 0; i < 5; i++) {
+                            try {
+                                userDAO.addUser(testUser);
```
Comment (andamian) [inline cadc-access-control-server/src/test/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAOTest.java]: It appears to me that the testUser are all identical (with or without HttpPrincipal) which is not exactly the main scenario of one real user with multiple duplicates that don't have HttpPrincipal

File: cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java (hunk 2 above)
Comment (hjeeves) [reply cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java]: I hear you. This was at Pat's suggestion. It was something that isn't supposed to be documented or available publicly. It's used only in testing.

File: cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java [hunk 4]
```diff
@@ -485,37 +495,112 @@ public User getUserRequest(final Principal userID)
...
+
+            // 'Real' user will have an LDAP_USER_NAME that is NOT EXTERNAL_USER_CN
+            if (EXTERNAL_USER_CN.equals(username)) {
+                continue;
+            }
+            ret = next;
+            break;
+        }
+
+        if (ret == null) {
```
[the remark this answers is not in the document]
Comment (hjeeves) [reply cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java]: Good point. Thanks. :)

File: cadc-access-control-server/src/test/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAOTest.java (hunk 3 above)
Comment (hjeeves) [reply cadc-access-control-server/src/test/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAOTest.java]: The key thing is userDAO.addUser is the internal method (inserts user directly into final tree, with $EXTERNAL-CN as LDAP-USER-NAME. the 'addUserRequest/approveUserRequest' cycle inserts the user information correctly into the ldap entry. HttpPrincipal existing or not isn't part of the check.

File: cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java [hunk 5]
```diff
@@ -525,61 +610,82 @@ private User getUser(final Principal userID, final String usersDN)
...
                 logger.debug(msg);
                 throw new UserNotFoundException(msg);
             }
-        }
-        catch (LDAPException e)
+
+            // Determine which is the 'real' user (not automatically generated
+            // by vospace interaction, for example,) and return that.
+            foundUser =  getUserFromResultList(multiSearchResult);
+            logger.debug("getUser: found " + userID.getName() + " in " + usersDN);
```
[the remark this answers is not in the document]
Comment (hjeeves) [reply cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java]: okay

File: cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java [hunk 6]
```diff
@@ -485,37 +495,112 @@ public User getUserRequest(final Principal userID)
...
+
+        if (StringUtil.hasLength(firstName) && StringUtil.hasLength(lastName)) {
+            newUser.personalDetails = new PersonalDetails(firstName, lastName);
+            newUser.personalDetails.address = userEntry.getAttributeValue(LDAP_ADDRESS);
+            newUser.personalDetails.city = userEntry.getAttributeValue(LDAP_CITY);
+            newUser.personalDetails.country = userEntry.getAttributeValue(LDAP_COUNTRY);
+            newUser.personalDetails.email = userEntry.getAttributeValue(LDAP_EMAIL);
+            newUser.personalDetails.institute = userEntry.getAttributeValue(LDAP_INSTITUTE);
+        }
+
```
[the remark this answers is not in the document]
Comment (hjeeves) [reply cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java]: okay

File: cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java (hunk 1 above)
Comment (hjeeves) [reply cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java]: Another good point. This particular function is part of the test code (and might get percolated to the cadc-access-control-admin command line interface at a later step.) It's not the only one that will detect multiple ldap entries. There isn't an explicit message blurted if more than one is found (unless it's flagged as an error.) - I think that's a ball drop on my part.  For the sake of tracking the error, I think an info level would be most likely to be seen. ?

File: cadc-access-control-server/src/test/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAOTest.java (hunk 3 above)
Comment (andamian) [reply cadc-access-control-server/src/test/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAOTest.java]: OK, I see. But I would say it's a bit risky to have the method you are testing (getUser) part of the setup of the test (line 889).

File: cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java (hunk 2 above)
Comment (andamian) [reply cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java]: Not documented but available publicly. Why is the initial value "cadc.skip.checkuser" and not just null? Does it have a meaning?

File: cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java [hunk 7]
```diff
@@ -507,6 +507,10 @@ public User getUserRequest(final Principal userID)
     private SearchResultEntry getFirstUserEntry(SearchResult multiSearchResult) {
         SearchResultEntry ret = null;
 
+        if (multiSearchResult == null) {
```
Comment (andamian) [inline cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/ldap/LdapUserDAO.java]: how about size()=0?

**Pull request merged**